plan/physical.go

```diff
@@ -308,9 +361,52 @@ func (ppn *PhysicalPlanNode) Cost(inStats []Statistics) (cost Cost, outStats Sta
 	return ppn.Spec.Cost(inStats)
 }
 
-// PhysicalAttributes encapsulates sny physical attributes of the result produced
+type PhysicalAttr interface {
+	SuccessorsMustRequire() bool
+}
+
+// PhysicalAttributes encapsulates any physical attributes of the result produced
 // by a physical plan node, such as collation, etc.
-type PhysicalAttributes struct {
+type PhysicalAttributes map[string]PhysicalAttr
+
+func (ppn *PhysicalPlanNode) SetOutputAttr(name string, v PhysicalAttr) {
+	if ppn.OutputAttrs == nil {
+		ppn.OutputAttrs = make(PhysicalAttributes)
+	}
+	ppn.OutputAttrs[name] = v
+}
+
+func (ppn *PhysicalPlanNode) SetRequiredAttr(name string, v PhysicalAttr) {
+	if ppn.RequiredAttrs == nil {
+		ppn.RequiredAttrs = make(PhysicalAttributes)
+	}
+	ppn.RequiredAttrs[name] = v
+}
+
+// Physical attributes used in specifying parallelization. The Run attribute
+// means the node executes in parallel. It accepts parallel data (a subset of
+// the source) and produces parallel data. The merge attribute means that the
+// node accepts parallel data, merges the streams, and produces non-parallel
+// data that covers the entire data source.
+const ParallelRunKey = "parallel-run"
+const ParallelMergeKey = "parallel-merge"
+
+type ParallelRunAttribute struct {
+	Factor int
+}
+
+// If a node produces parallel data, then all successors must require parallel
+// data, otherwise there will be a plan error.
+func (ParallelRunAttribute) SuccessorsMustRequire() bool {
+	return true
+}
+
+type ParallelMergeAttribute struct {
+	Factor int
+}
+
+func (ParallelMergeAttribute) SuccessorsMustRequire() bool {
+	return false
 }
```

Comment:
It might be good if all the attributes were in a file on their own?

Reply:
Done.